web3_proxy/src/frontend/users/referral.rs

```diff
@@ -70,3 +75,132 @@ pub async fn user_referral_link_get(
+) -> Web3ProxyResponse {
+    // First get the bearer token and check if the user is logged in
+    let (user, _semaphore) = app.bearer_is_authorized(bearer).await?;
+
+    let db_replica = app
+        .db_replica()
+        .context("getting replica db for user's revert logs")?;
+
+    // Get all referral records associated with this user
+    let referrals = referee::Entity::find()
+        .filter(referee::Column::UserId.eq(user.id))
+        .find_also_related(referrer::Entity)
+        .all(db_replica.conn())
+        .await?;
+
+    // For each related referral person, find the corresponding user-address
+    #[derive(Debug, Serialize)]
+    struct Info {
+        credits_applied_for_referee: bool,
+        credits_applied_for_referrer: Decimal, // Probably make this a float instead (approximate anyways)
+        referral_start_date: DateTime,
+        used_referral_code: String,
+    };
+
+    let mut out: Vec<Info> = Vec::new();
+    for x in referrals.into_iter() {
+        let (referral_record, referrer_record) = (x.0, x.1.unwrap());
+        // The foreign key is never optional
+        let referring_user = user::Entity::find_by_id(referrer_record.user_id)
+            .one(db_replica.conn())
+            .await?
+            .context("Database error, no foreign key found for referring user")?;
+        let tmp = Info {
+            credits_applied_for_referee: referral_record.credits_applied_for_referee,
+            credits_applied_for_referrer: referral_record.credits_applied_for_referrer,
+            referral_start_date: referral_record.referral_start_date,
+            used_referral_code: referrer_record.referral_code,
+        };
+        // Start inserting json's into this
+        out.push(tmp);
+    }
+
+    // Turn this into a response
+    let response_json = json!({
+        "referrals": out,
+        "user": user,
+    });
+
+    let response = (StatusCode::OK, Json(response_json)).into_response();
+    Ok(response)
+}
+
+#[debug_handler]
+pub async fn user_shared_referral_stats(
+    Extension(app): Extension<Arc<Web3ProxyApp>>,
+    TypedHeader(Authorization(bearer)): TypedHeader<Authorization<Bearer>>,
+    Query(_params): Query<HashMap<String, String>>,
+) -> Web3ProxyResponse {
+    // First get the bearer token and check if the user is logged in
+    let (user, _semaphore) = app.bearer_is_authorized(bearer).await?;
+
+    let db_replica = app
+        .db_replica()
+        .context("getting replica db for user's revert logs")?;
+
+    // Get all referral records associated with this user
+    let referrals = referrer::Entity::find()
+        .filter(referrer::Column::UserId.eq(user.id))
+        .find_also_related(referee::Entity)
+        .all(db_replica.conn())
+        .await?;
+
+    // Return early if the user does not have any referred entities
+    if referrals.len() == 0 {
+        let response_json = json!({
+        "referrals": Vec::<()>::new(),
+        "used_referral_code": "",
```

Comment:
do we want `""`, or would it be better to do `None` (which json will turn into `null`)?